File: runtimes/python/analyzer.py
"""
Python AST Analyzer for MCP Orchestrator
Analyzes Python code and extracts functions, inputs, outputs, complexity, branches
"""

import ast
import sys
import json
from typing import List, Dict, Any, Set, Optional

class PythonCodeAnalyzer(ast.NodeVisitor):
    def __init__(self):
        self.functions = []
        self.current_function = None
        self.total_branches = 0
        self.complexity = 0
        self.side_effects = set()
        
    def analyze(self, code: str) -> Dict[str, Any]:
        """Analisa o código Python e retorna métricas estruturadas"""
        try:
            tree = ast.parse(code)
            self.visit(tree)
            
            return {
                "functions": self.functions,
                "inputs": self._collect_all_inputs(),
                "outputs": self._collect_all_outputs(), 
                "complexity": self.complexity,
                "branches": self.total_branches,
                "side_effects": list(self.side_effects)
            }
        except SyntaxError as e:
            return {"error": f"Syntax error: {e}"}
        except Exception as e:
            return {"error": f"Analysis error: {e}"}
    
    def visit_FunctionDef(self, node: ast.FunctionDef):
        """Visita definições de função"""
        func_info = {
            "name": node.name,
            "inputs": self._extract_function_inputs(node),
            "outputs": self._extract_function_outputs(node),
            "line_start": node.lineno,
            "line_end": node.end_lineno if hasattr(node, 'end_lineno') else node.lineno,
            "branches": 0,
            "local_complexity": 1  # Base complexity
        }
        
        # Salvar contexto atual
        prev_function = self.current_function
        self.current_function = func_info
        
        # Analisar corpo da função
        for stmt in node.body:
            self.visit(stmt)
            
        # Atualizar métricas da função
        func_info["branches"] = self.current_function["branches"] 
        func_info["local_complexity"] = self.current_function["local_complexity"]
        
        self.functions.append(func_info)
        self.total_branches += func_info["branches"]
        self.complexity += func_info["local_complexity"]
        
        # Restaurar contexto
        self.current_function = prev_function
        
        # Não visitar filhos automaticamente (já visitamos manualmente)
    
    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
        """Visita funções assíncronas"""
        self.visit_FunctionDef(node)  # Mesmo tratamento
        self.side_effects.add("async_operations")
    
    def visit_If(self, node: ast.If):
        """Conta branches condicionais"""
        if self.current_function:
            self.current_function["branches"] += 1
            self.current_function["local_complexity"] += 1
            
        self.generic_visit(node)
    
    def visit_For(self, node: ast.For):
        """Conta loops como branches"""
        if self.current_function:
            self.current_function["branches"] += 1
            self.current_function["local_complexity"] += 1
            
        self.generic_visit(node)
    
    def visit_While(self, node: ast.While):
        """Conta loops while"""
        if self.current_function:
            self.current_function["branches"] += 1
            self.current_function["local_complexity"] += 1
            
        self.generic_visit(node)
    
    def visit_Try(self, node: ast.Try):
        """Conta try/except como branches"""
        if self.current_function:
            # Try + cada except/else/finally
            branches_count = 1 + len(node.handlers)
            if node.orelse:
                branches_count += 1
            if node.finalbody:
                branches_count += 1
                
            self.current_function["branches"] += branches_count
            self.current_function["local_complexity"] += branches_count
            
        self.generic_visit(node)
    
    def visit_With(self, node: ast.With):
        """Context managers podem ter side effects"""
        self.side_effects.add("context_managers")
        self.generic_visit(node)
    
    def visit_Call(self, node: ast.Call):
        """Detecta chamadas que podem ter side effects"""
        # Heurísticas para detectar side effects
        if isinstance(node.func, ast.Name):
            func_name = node.func.id
            if func_name in ['print', 'open', 'input', 'exec', 'eval']:
                self.side_effects.add("io_operations")
            elif func_name in ['setattr', 'delattr', 'globals', 'locals']:
                self.side_effects.add("state_modification")
                
        elif isinstance(node.func, ast.Attribute):
            attr_name = node.func.attr
            if attr_name in ['write', 'read', 'append', 'close', 'flush']:
                self.side_effects.add("io_operations") 
            elif attr_name in ['send', 'get', 'post', 'put', 'delete']:
                self.side_effects.add("network_operations")
            elif attr_name in ['commit', 'rollback', 'execute']:
                self.side_effects.add("database_operations")
                
        self.generic_visit(node)
    
    def visit_Global(self, node: ast.Global):
        """Uso de variáveis globais"""
        self.side_effects.add("global_state")
        self.generic_visit(node)
        
    def visit_Nonlocal(self, node: ast.Nonlocal):
        """Uso de variáveis não-locais"""
        self.side_effects.add("nonlocal_state")
        self.generic_visit(node)
    
    def _extract_function_inputs(self, node: ast.FunctionDef) -> List[Dict[str, Any]]:
        """Extrai parâmetros da função"""
        inputs = []
        
        # Argumentos posicionais
        for arg in node.args.args:
            param = {
                "name": arg.arg,
                "type": self._get_annotation(arg.annotation) if arg.annotation else "Any",
                "kind": "positional"
            }
            inputs.append(param)
        
        # Argumentos *args
        if node.args.vararg:
            inputs.append({
                "name": f"*{node.args.vararg.arg}",
                "type": self._get_annotation(node.args.vararg.annotation) if node.args.vararg.annotation else "tuple",
                "kind": "varargs"
            })
            
        # Argumentos **kwargs
        if node.args.kwarg:
            inputs.append({
                "name": f"**{node.args.kwarg.arg}",
                "type": self._get_annotation(node.args.kwarg.annotation) if node.args.kwarg.annotation else "dict", 
                "kind": "kwargs"
            })
        
        # Keyword-only arguments
        for arg in node.args.kwonlyargs:
            param = {
                "name": arg.arg,
                "type": self._get_annotation(arg.annotation) if arg.annotation else "Any",
                "kind": "keyword_only"
            }
            inputs.append(param)
            
        return inputs
    
    def _extract_function_outputs(self, node: ast.FunctionDef) -> List[str]:
        """Extrai tipo de retorno da função"""
        outputs = []
        
        # Type annotation no return
        if node.returns:
            return_type = self._get_annotation(node.returns)
            outputs.append(return_type)
        else:
            # Tentar inferir do corpo da função
            return_types = set()
            for stmt in ast.walk(node):
                if isinstance(stmt, ast.Return) and stmt.value:
                    inferred = self._infer_return_type(stmt.value)
                    if inferred:
                        return_types.add(inferred)
            
            outputs.extend(list(return_types) if return_types else ["Any"])
            
        return outputs
    
    def _get_annotation(self, annotation: Optional[ast.AST]) -> str:
        """Converte annotation AST para string"""
        if annotation is None:
            return "Any"
            
        if isinstance(annotation, ast.Name):
            return annotation.id
        elif isinstance(annotation, ast.Constant):
            return str(annotation.value) 
        elif isinstance(annotation, ast.Attribute):
            return f"{self._get_annotation(annotation.value)}.{annotation.attr}"
        elif isinstance(annotation, ast.Subscript):
            value = self._get_annotation(annotation.value)
            slice_val = self._get_annotation(annotation.slice)
            return f"{value}[{slice_val}]"
        else:
            return "Complex"
    
    def _infer_return_type(self, return_value: ast.AST) -> Optional[str]:
        """Infere tipo de retorno baseado no valor"""
        if isinstance(return_value, ast.Constant):
            if isinstance(return_value.value, bool):
                return "bool"
            elif isinstance(return_value.value, int):
                return "int"
            elif isinstance(return_value.value, float):
                return "float"
            elif isinstance(return_value.value, str):
                return "str"
            elif return_value.value is None:
                return "None"
        elif isinstance(return_value, ast.List):
            return "list"
        elif isinstance(return_value, ast.Dict):
            return "dict"
        elif isinstance(return_value, ast.Tuple):
            return "tuple"
        elif isinstance(return_value, ast.Set):
            return "set"
            
        return None
    
    def _collect_all_inputs(self) -> List[str]:
        """Coleta todos os inputs de todas as funções"""
        all_inputs = []
        for func in self.functions:
            for inp in func["inputs"]:
                param_str = f"{inp['name']}: {inp['type']}"
                if param_str not in all_inputs:
                    all_inputs.append(param_str)
        return all_inputs
    
    def _collect_all_outputs(self) -> List[str]:
        """Coleta todos os outputs de todas as funções"""
        all_outputs = []
        for func in self.functions:
            for out in func["outputs"]:
                if out not in all_outputs:
                    all_outputs.append(out)
        return all_outputs

def main():
    """Entry point - lê código do stdin e imprime análise JSON"""
    try:
        # Ler código do stdin
        code = sys.stdin.read()
        if not code.strip():
            print(json.dumps({"error": "No code provided"}))
            return
        
        # Analisar código
        analyzer = PythonCodeAnalyzer()
        result = analyzer.analyze(code)
        
        # Imprimir resultado como JSON
        print(json.dumps(result, indent=None, separators=(',', ':')))
        
    except Exception as e:
        error_result = {"error": f"Analyzer failure: {str(e)}"}
        print(json.dumps(error_result))
        sys.exit(1)

if __name__ == "__main__":
    main()

File: runtimes/ts/analyzer.js
#!/usr/bin/env node
/**
 * TypeScript/JavaScript AST Analyzer for MCP Orchestrator
 * Analyzes TypeScript/JavaScript code and extracts functions, inputs, outputs, complexity, branches
 */

const { Project, SyntaxKind } = require('ts-morph');

class TypeScriptAnalyzer {
    constructor() {
        this.functions = [];
        this.totalBranches = 0;
        this.complexity = 0;
        this.sideEffects = new Set();
    }

    analyze(code) {
        try {
            const project = new Project({ useInMemoryFileSystem: true });
            const sourceFile = project.createSourceFile('temp.ts', code);
            
            this.visitNode(sourceFile);
            
            return {
                functions: this.functions,
                inputs: this.collectAllInputs(),
                outputs: this.collectAllOutputs(),
                complexity: this.complexity,
                branches: this.totalBranches,
                side_effects: Array.from(this.sideEffects)
            };
        } catch (error) {
            return { error: `Analysis error: ${error.message}` };
        }
    }

    visitNode(node) {
        if (node.getKind() === SyntaxKind.FunctionDeclaration ||
            node.getKind() === SyntaxKind.ArrowFunction ||
            node.getKind() === SyntaxKind.FunctionExpression ||
            node.getKind() === SyntaxKind.MethodDeclaration) {
            
            this.analyzeFunctionNode(node);
        }

        // Visit conditional statements
        if (node.getKind() === SyntaxKind.IfStatement) {
            this.totalBranches++;
            this.complexity++;
        }

        // Visit loops
        if (node.getKind() === SyntaxKind.ForStatement ||
            node.getKind() === SyntaxKind.WhileStatement ||
            node.getKind() === SyntaxKind.DoStatement ||
            node.getKind() === SyntaxKind.ForInStatement ||
            node.getKind() === SyntaxKind.ForOfStatement) {
            this.totalBranches++;
            this.complexity++;
        }

        // Visit try-catch
        if (node.getKind() === SyntaxKind.TryStatement) {
            this.totalBranches += 2; // try + catch
            this.complexity += 2;
        }

        // Visit switch statements
        if (node.getKind() === SyntaxKind.SwitchStatement) {
            const caseCount = node.getCaseBlock().getClauses().length;
            this.totalBranches += Math.max(1, caseCount);
            this.complexity += Math.max(1, caseCount);
        }

        // Detect side effects
        if (node.getKind() === SyntaxKind.CallExpression) {
            this.detectSideEffects(node);
        }

        // Continue visiting children
        node.forEachChild(child => this.visitNode(child));
    }

    analyzeFunctionNode(node) {
        const funcInfo = {
            name: this.getFunctionName(node),
            inputs: this.extractInputs(node),
            outputs: this.extractOutputs(node),
            line_start: node.getStartLineNumber(),
            line_end: node.getEndLineNumber(),
            branches: 0,
            local_complexity: 1
        };

        // Count branches within this function
        const currentBranches = this.totalBranches;
        const currentComplexity = this.complexity;
        
        node.forEachChild(child => this.visitNode(child));
        
        funcInfo.branches = this.totalBranches - currentBranches;
        funcInfo.local_complexity = this.complexity - currentComplexity + 1;

        this.functions.push(funcInfo);
    }

    getFunctionName(node) {
        if (node.getKind() === SyntaxKind.FunctionDeclaration) {
            return node.getName() || '<anonymous>';
        }
        if (node.getKind() === SyntaxKind.MethodDeclaration) {
            return node.getName();
        }
        if (node.getKind() === SyntaxKind.ArrowFunction || 
            node.getKind() === SyntaxKind.FunctionExpression) {
            // Try to get name from variable assignment
            const parent = node.getParent();
            if (parent && parent.getKind() === SyntaxKind.VariableDeclaration) {
                return parent.getName();
            }
            return '<anonymous>';
        }
        return '<unknown>';
    }

    extractInputs(node) {
        const inputs = [];
        
        if (node.getParameters) {
            const parameters = node.getParameters();
            for (const param of parameters) {
                inputs.push({
                    name: param.getName(),
                    type: param.getTypeNode() ? param.getTypeNode().getText() : 'any',
                    kind: 'parameter',
                    optional: param.hasQuestionToken()
                });
            }
        }
        
        return inputs;
    }

    extractOutputs(node) {
        const outputs = [];
        
        if (node.getReturnTypeNode) {
            const returnType = node.getReturnTypeNode();
            if (returnType) {
                outputs.push(returnType.getText());
            }
        }
        
        if (outputs.length === 0) {
            // Try to infer from return statements
            const returnStatements = node.getDescendantsOfKind(SyntaxKind.ReturnStatement);
            if (returnStatements.length > 0) {
                outputs.push('inferred');
            } else {
                outputs.push('void');
            }
        }
        
        return outputs;
    }

    detectSideEffects(callNode) {
        const expression = callNode.getExpression();
        
        if (expression.getKind() === SyntaxKind.Identifier) {
            const funcName = expression.getText();
            if (['console.log', 'alert', 'confirm', 'prompt'].includes(funcName)) {
                this.sideEffects.add('io_operations');
            }
        }
        
        if (expression.getKind() === SyntaxKind.PropertyAccessExpression) {
            const propAccess = expression;
            const propertyName = propAccess.getName();
            
            if (['fetch', 'XMLHttpRequest', 'axios'].includes(propertyName) ||
                propAccess.getText().includes('.fetch(') ||
                propAccess.getText().includes('.post(') ||
                propAccess.getText().includes('.get(')) {
                this.sideEffects.add('network_operations');
            }
            
            if (['localStorage', 'sessionStorage', 'indexedDB'].includes(propertyName)) {
                this.sideEffects.add('storage_operations');
            }
            
            if (propertyName === 'log' && propAccess.getExpression().getText() === 'console') {
                this.sideEffects.add('io_operations');
            }
        }
        
        // Detect async operations
        if (callNode.getParent() && callNode.getParent().getKind() === SyntaxKind.AwaitExpression) {
            this.sideEffects.add('async_operations');
        }
    }

    collectAllInputs() {
        const allInputs = [];
        for (const func of this.functions) {
            allInputs.push(...func.inputs.map(input => `${input.name}: ${input.type}`));
        }
        return allInputs;
    }

    collectAllOutputs() {
        const allOutputs = [];
        for (const func of this.functions) {
            allOutputs.push(...func.outputs);
        }
        return allOutputs;
    }
}

// Main execution
if (require.main === module) {
    let input = '';
    
    process.stdin.setEncoding('utf8');
    process.stdin.on('readable', () => {
        const chunk = process.stdin.read();
        if (chunk !== null) {
            input += chunk;
        }
    });
    
    process.stdin.on('end', () => {
        const analyzer = new TypeScriptAnalyzer();
        const result = analyzer.analyze(input);
        console.log(JSON.stringify(result, null, 2));
    });
}

module.exports = TypeScriptAnalyzer;